### PR TITLE
Integrate privacy into field and method selection

### LINKED ISSUE #12808: Integrate privacy into autoderef loop with overloaded deref
Once PR #12610 lands, the autoderef loop will be generalized to permit dereferencing through user-defined structures that implement the `Deref` trait. As such, the termination condition becomes less clear around privacy. Currently, when searching for fields, we autoderef until finding a struct and then search for a field. In PR #12610, we will continue searching if no field with the desired name is found. Unfortunately, the PR does not consider privacy, and thus adding a field to a smart pointer can shadow fields in the referent. The same applies to methods. This is suboptimal.

The desired behavior is that private fields and methods are ignored as part of the autoderef loop, so long as the expression being autoderef'd is not part of the module where the field/method was defined. This is a relatively simple check: if the field/method is declared private, you find the module M containing the declaration. You then check the whether current module N is a equal to M or a submodule of M. If not, you ignore the field/method.

For better error messages, it would be nice to at least track *whether* a private field/method with a suitable name was seen, so that we can report an error like "no accessible field `f` found (note that private fields were found)" or something like that.

### LINKED ISSUE #22684: A strange processing methods visibility
rustc 1.0.0-nightly (2b01a37ec 2015-02-21) (built 2015-02-22)

I create file with content:
```
use std::hash::SipHasher;
use std::hash::Hasher;

fn main() {
}

fn compile_fail() {
	let hasher = SipHasher::new();
	hasher.write(&[0, 1, 2]);
}

fn compile_work() {
	let sip_hasher = SipHasher::new();
	let hasher: &mut Hasher = &mut sip_hasher;
	hasher.write(&[0, 1, 2]);
}
```

And have compilation errors:
```
private.rs:9:2: 9:26 error: method `write` is private
private.rs:9 	hasher.write(&[0, 1, 2]);
             	^~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

I do not understand what is happening due to this error, because this cast is essentially doing nothing.

### PR DESCRIPTION
**Pull request by jseyfried**
This PR integrates privacy checking into field and method selection so that an inaccessible field/method cannot stop an accessible field/method from being used (fixes #12808 and fixes #22684).
r? @eddyb

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @nikomatsakis @petrochenkov

**Comment by nikomatsakis** [comment]
On Sat, Feb 27, 2016 at 01:23:08AM -0800, Jeffrey Seyfried wrote:
> cc @nikomatsakis @petrochenkov 

you're on a roll! Just FYI I sat down over last few days to read over
some of your changes to resolve in detail, because I realize I've kind
of lost track of how things are working, and am starting to feel
uncomfortable reviewing :) hoping to finish that up tonight though

**Comment by jseyfried** [comment]
@nikomatsakis Ok, let me know if you like me to explain or comment anything!

**Comment by jseyfried** [comment]
Fixes #22684

**Comment by eddyb** [comment]
@jseyfried you need to put that in the description.

**Comment by jseyfried** [comment]
@eddy I updated the description.

**Comment by jseyfried** [comment]
I rebased and added a commit that refactors away most of the privacy visitor.

**Comment by nikomatsakis** [comment]
@jseyfried can you add some tests that show that the logic about private fields/methods is working? I was thinking it'd be nice to have a test where the private field appears in an outer deref. In other words, if the autoderef chain of types is `A -> B -> C`, then the field/method is on type A or B.

r=me with such a test.

File: src/test/compile-fail/autoderef-privacy.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    impl ::std::ops::Deref for Baz {
+        type Target = ::Baz2;
+        fn deref(&self) -> &::Baz2 { unimplemented!() }
+    }
+}
+
+fn f(bar: foo::Bar, baz: foo::Baz) {
+    let _ = bar.i;
+    let _ = baz.0;
+    let _ = bar.f();
```
Comment (nikomatsakis) [inline src/test/compile-fail/autoderef-privacy.rs]: can you convert this to a run-pass test and somehow check the return value of `f`, so that we know the right function is being called?

File: src/test/compile-fail/issue-22684.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        fn bar(&self) -> bool {}
+    }
+    impl Baz for Foo {}
+}
+
+fn main() {
+    use foo::Baz;
+
+    // Check that `bar` resolves to the trait method, not the inherent impl method.
+    let _: () = foo::Foo.bar(); //~ ERROR mismatched types
```
Comment (nikomatsakis) [inline src/test/compile-fail/issue-22684.rs]: this is good

File: src/test/compile-fail/autoderef-privacy.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    impl ::std::ops::Deref for Baz {
+        type Target = ::Baz2;
+        fn deref(&self) -> &::Baz2 { unimplemented!() }
+    }
+}
+
+fn f(bar: foo::Bar, baz: foo::Baz) {
+    let _ = bar.i;
```
Comment (nikomatsakis) [inline src/test/compile-fail/autoderef-privacy.rs]: (similarly, in this run-pass test, check that the correct field is being selected)

**Comment by nikomatsakis** [comment]
Sorry, let me clarify my previous comment. I'd like to see two things:

1. Modify the existing test for private fields/methods to be run-pass, and using some logic that proves we are picking the right fields.
    - Maybe also extend the test to read from the fields in a context where it should pick the private one, and prove that *this*  logic is working correctly.
2. A test specifically targeting the case where only a private field/method is available, which would target the [logic you put in around that case](https://github.com/rust-lang/rust/pull/31938/commits/8afb0802e401620680f96592e62342f5f3f2e8c9#diff-1d1b0d29a2e8da97c6bfb6e364d920c7R2971).

File: src/test/compile-fail/autoderef-privacy.rs (hunk 1 above)
Comment (jseyfried) [inline src/test/compile-fail/autoderef-privacy.rs]: Will do

**Comment by jseyfried** [comment]
@nikomatsakis 
>
1. Modify the existing test for private fields/methods to be run-pass, and using some logic that proves we are picking the right fields.
 - Maybe also extend the test to read from the fields in a context where it should pick the private one, and prove that this logic is working correctly.

Done.

> A test specifically targeting the case where only a private field/method is available

Private inherent methods are tested in `private-method-inherited` (and others) and I amended `struct-field-privacy` to test private positional fields as well as private named fields.

I believe these tests are sufficient to check the logic you referenced, but if you want I can add a `compile-fail` test where there is only a private field/method in a non-trivial autoderef chain.

**Comment by nikomatsakis** [comment]
@jseyfried I'll leave it to your judgement. I certainly lean towards more exhaustive testing but there is no reason to add duplicate tests either. R=me either way.

**Comment by jseyfried** [comment]
Ok, thanks for reviewing!

@bors r=nikomatsakis

**Comment by alexcrichton** [comment]
@bors: retry force clean

maybe second time's the charm?
